andela/knights-bn-backend
# #170947660 Implement Facebook and Google login

#### What does this PR do?
 This PR helps a user to login with social networks: Facebook and Google.
#### Description of Task to be completed?
- Create  barefoot-nomad  API on Google cloud platform.
- Create barefoot-nomad app on Facebook for developers platform.
- Install passport package as an authentication middleware
- Install passport-google-oauth20 package to authenticate with Google.
- Install passport-facebook package to authenticate with Facebook.
- Connect user to DB
- Write tests
- Write documentation

#### How should this be manually tested?
- run `git clone https://github.com/andela/knights-bn-backend.git` to clone the repo.
- run `cd knights-bn-backend` to go to the project directory.
- run `git checkout ft-config-oauth-170947660` to switch to this branch.
- run `npm install` to install dependencies.
- create a Postgres database locally.
- create a .env file then add environment variable as shown below:
```
DB_USER='name of the user of the database'
DB_PASS='password of your Postgres'
DB_NAME='name of the database || Barefoot'
DB_TEST_NAME='name of the database for test || Barefoot_test'
NODE_ENV='development'
SECRETKEY='PrivateKey'
GOOGLE_CLIENT_ID = 'XXXXX'
GOOGLE_CLIENT_SECRET = 'XXXXX'
GOOGLE_CALLBACK_URL = 'XXXXX'
FCBK_CLIENT_ID = 'XXXXX'
FCBK_CLIENT_SECRET = 'XXXXX'
FCBK_CALLBACK_URL = 'XXXXX'

```
- run `npm run createTables` to create the database tables
- run `npm start` to start the server.
- Open your browser and copy this URL `http://localhost:4000/api/v1/auth/login/socialLogin`
- you should see a page with two buttons `Login with Google` and `Login with Facebook` fill in your credentials and follow the steps.


#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[link](https://www.pivotaltracker.com/story/show/170947660)

#### Screenshots

### `Login with Google`

### 1. social login page
![image](https://user-images.githubusercontent.com/30593186/74308296-7acaac00-4d70-11ea-85d8-09a4d6ceb886.png)
### 2. Google consent screen
![image](https://user-images.githubusercontent.com/30593186/74325407-36033d00-4d91-11ea-85c9-12ca162b73f7.png)
### 3. Retrieved user info
![image](https://user-images.githubusercontent.com/30593186/74325119-c55c2080-4d90-11ea-9086-b4f5193e47d7.png)

### `Login with Facebook`

### 1. Fill-in the credentials
![image](https://user-images.githubusercontent.com/30593186/74368622-b307d400-4ddc-11ea-85dc-6af536fc0407.png)
### 2. Get retrieved user info
![image](https://user-images.githubusercontent.com/30593186/74368901-2578b400-4ddd-11ea-9f37-17f3aa4ca1b7.png)
